Fix generated primes being too large

colatkinson opened the pull request:
Simple fix for #124, simply by testing if the prime is too big and recurring if so.

Also changes prime generation to search from [2^(n-1), 2^n) instead of [2^(n-1) + 2^(n-2), 2^n] by setting only the top bit, as is the general standard (see e.g. [OpenSSL](https://github.com/openssl/openssl/blob/master/crypto/bn/bn_prime.c#L101)).

colatkinson committed:
Fix generate(Safe)Prime to guarantee prime size

Add check for size in generatePrime

Add size test in generateSafePrime

Require only that top bit is set, instead of top 2

This is the general standard, see e.g. OpenSSL

Add an error for too few bits being supplied to prime generator, and add documentation

Add some documentation and require highest two bits set

Simplify return syntax in generatePrime and generateSafePrime

Switch exponent to bit-shift for small performance boost

colatkinson committed:
Fix bug in isProbablyPrime for small numbers

Fix bug in isProbablyPrime where too many iterations were specified for numbers less than 100

Add clause to isProbablyPrime to use hardcoded values <= 2903

ocheron commented on Crypto/Number/Prime.hs:
```diff
@@ -37,25 +38,54 @@ import Data.Bits
     | otherwise                                            = False
+    where
+        halfn = n `div` 2
+        iters = fromInteger (min 50 halfn)
```
After testing small primes, a short path returning `True` when `n >= 2 && n <= 2903` would avoid calling `primalityTestFermat` completely.

ocheron commented on Crypto/Number/Prime.hs:
```diff
@@ -37,25 +38,54 @@ import Data.Bits
+            return $ prime
+        else do
+            ret <- generatePrime bits
+            return $ ret
```
Using `numBits` from `Crypto.Number.Basic` to test the bit size would probably be more efficient (with GMP at least).

Also
```haskell
else do ret <- generatePrime bits
        return $ ret
```
can be shorten to:
```haskell
else generatePrime bits
```

colatkinson committed:
Update tests for new generate(Safe)Prime

Update generatePrime test to test smaller bit sizes

Add test for generateSafePrime

Remove -fobject-code
